### DISCUSSION TITLE
Can I use the websocket context with the regular context?

### DISCUSSION THREAD

**Discussion by thejessewinton** [trpc/trpc#3509] [Q&A] [answered]
This might be a really dumb question, but am I able to use the `websocket/subscription` API alongside a traditional context? So, could I use them both to be able to have certain routers that utilize traditional queries and mutations, and other routers that use the websocket server? If so, is there an example somewhere of how to implement it?

Thanks in advance!

**Comment by KATT** [comment, marked as answer]
See

- https://github.com/trpc/examples-next-prisma-websockets-starter for an example using WSS for all client-side comms but can still do SSR through normal http calls
- https://trpc.io/docs/links/splitLink for dynamically deciding if a call should be http or wss

**Reply by thejessewinton** [reply]
Thank you! Can’t say enough how awesome tRPC is as well, thanks for the work you’ve done for the community!

**Comment by AmadoGuerrero** [comment]
I might be using splitLink wrong or something, but I'm setting a condition like `op.type === 'subscription'` to determine if I should be using a ws link, or httpBatchLink, and I FEEL like it should be working. But whenver I try that method, my websocket calls work fine, but my regular httpBatchLinks (I also tried httpLink) calls hang and I don't get a response from the server. 

`API resolved without sending a response for /api/trpc/<pathtoroute>, this may result in stalled requests.`

If I use wsLink for all requests everything works fine, but I would like to make it so that only my ws links go through the ws server.
I'm using the t3-stack starter which comes with a few built-in helper functions and wrappers, so its possible that might be where my issue lies.

I have a good feeling my problem is very closely related to this question, but if I need to make a new thread, I will.

Whats interesting is that if I throw a console log right before creating the context, I see the log in the ws server, but not in the regular server.

Here are the relevant files:

wsServer.ts
```typescript
import ws from "ws";
import { applyWSSHandler } from "@trpc/server/adapters/ws";
import { createServer } from "https";
import { appRouter } from "~/server/api/root";
import { createTRPCContext } from "~/server/api/trpc";
import { readFileSync } from "fs";

const server = createServer({
  key: readFileSync('./key.pem'),
  cert: readFileSync('./cert.pem'),
});

const port = 5000;
const wss = new ws.Server({
  server,
});

const handler = applyWSSHandler({ wss, createContext: createTRPCContext, router: appRouter });

wss.on("connection", (ws) => {
  console.log(`++ ws connection (${wss.clients.size})`);

  wss.once("close", () => {
    console.log(`-- ws connection (${wss.clients.size})`);
  });
});

console.log(`ws server started on port: ${port}`);

process.on("SIGTERM", () => {
  console.log(`SIGTERM`);

  handler.broadcastReconnectNotification();

  wss.close();
});

server.listen(port);
```

~/server/api/trpc
```typescript
/**
 * YOU PROBABLY DON'T NEED TO EDIT THIS FILE, UNLESS:
 * 1. You want to modify request context (see Part 1).
 * 2. You want to create a new middleware or type of procedure (see Part 3).
 *
 * TL;DR - This is where all the tRPC server stuff is created and plugged in. The pieces you will
 * need to use are documented accordingly near the end.
 */

/**
 * 1. CONTEXT
 *
 * This section defines the "contexts" that are available in the backend API.
 *
 * These allow you to access things when processing a request, like the database, the session, etc.
 */
import { type CreateNextContextOptions } from "@trpc/server/adapters/next";
import { type Session } from "next-auth";

import { getServerAuthSession } from "~/server/auth";
import { prisma, redis } from "~/server/db";
import type ws from "ws";

type CreateContextOptions = {
  session: Session | null;
};

/**
 * This helper generates the "internals" for a tRPC context. If you need to use it, you can export
 * it from here.
 *
 * Examples of things you may need it for:
 * - testing, so we don't have to mock Next.js' req/res
 * - tRPC's `createSSGHelpers`, where we don't have req/res
 *
 * @see https://create.t3.gg/en/usage/trpc#-servertrpccontextts
 */
const createInnerTRPCContext = (opts: CreateContextOptions) => {
  return {
    session: opts.session,
    prisma,
    redis,
  };
};

/**
 * This is the actual context you will use in your router. It will be used to process every request
 * that goes through your tRPC endpoint.
 *
 * @see https://trpc.io/docs/context
 */
export const createTRPCContext = async (
  opts: CreateNextContextOptions | NodeHTTPCreateContextFnOptions<IncomingMessage, ws>
) => {

  console.log(`Attempting to get the session from the server...`);

  // Get the session from the server using the getServerSession wrapper function
  const session = await getServerAuthSession(opts);

  console.log(`createTRPCContext for ${session?.user?.name ?? 'unknown user'}`);

  return createInnerTRPCContext({
    session,
  });
};

/**
 * 2. INITIALIZATION
 *
 * This is where the tRPC API is initialized, connecting the context and transformer.
 */
import { initTRPC, TRPCError } from "@trpc/server";
import superjson from "superjson";
import type { NodeHTTPCreateContextFnOptions } from "@trpc/server/dist/adapters/node-http/types";
import type { IncomingMessage } from "http";

const t = initTRPC.context<typeof createTRPCContext>().create({
  transformer: superjson,
  errorFormatter({ shape }) {
    return shape;
  },
});

/**
 * 3. ROUTER & PROCEDURE (THE IMPORTANT BIT)
 *
 * These are the pieces you use to build your tRPC API. You should import these a lot in the
 * "/src/server/api/routers" directory.
 */

/**
 * This is how you create new routers and sub-routers in your tRPC API.
 *
 * @see https://trpc.io/docs/router
 */
export const createTRPCRouter = t.router;

/**
 * Public (unauthenticated) procedure
 *
 * This is the base piece you use to build new queries and mutations on your tRPC API. It does not
 * guarantee that a user querying is authorized, but you can still access user session data if they
 * are logged in.
 */
export const publicProcedure = t.procedure;

/** Reusable middleware that enforces users are logged in before running the procedure. */
const enforceUserIsAuthed = t.middleware(({ ctx, next }) => {
  if (!ctx.session || !ctx.session.user) {
    throw new TRPCError({ code: "UNAUTHORIZED" });
  }
  return next({
    ctx: {
      // infers the `session` as non-nullable
      session: { ...ctx.session, user: ctx.session.user },
    },
  });
});

/**
 * Protected (authenticated) procedure
 *
 * If you want a query or mutation to ONLY be accessible to logged in users, use this. It verifies
 * the session is valid and guarantees `ctx.session.user` is not null.
 *
 * @see https://trpc.io/docs/procedures
 */
export const protectedProcedure = t.procedure.use(enforceUserIsAuthed);
```

~/server/auth
```typescript
import { type GetServerSidePropsContext } from "next";
import { getServerSession, type NextAuthOptions, type DefaultSession } from "next-auth";
import DiscordProvider from "next-auth/providers/discord";
import { PrismaAdapter } from "@next-auth/prisma-adapter";
import { env } from "~/env.mjs";
import { prisma } from "~/server/db";
import { getSession } from "next-auth/react";
import { type UserRole } from "@prisma/client";
import { CreateNextContextOptions } from "@trpc/server/adapters/next";
import { NodeHTTPCreateContextFnOptions } from "@trpc/server/dist/adapters/node-http";
import { IncomingMessage } from "http";
import ws from "ws";

/**
 * Module augmentation for `next-auth` types. Allows us to add custom properties to the `session`
 * object and keep type safety.
 *
 * @see https://next-auth.js.org/getting-started/typescript#module-augmentation
 */
declare module "next-auth" {
  interface Session extends DefaultSession {
    user: {
      id: string;
      role: UserRole;
    } & DefaultSession["user"];
  }

  // interface User {
  //   // ...other properties
  //   // role: UserRole;
  // }
}

/**
 * Options for NextAuth.js used to configure adapters, providers, callbacks, etc.
 *
 * @see https://next-auth.js.org/configuration/options
 */
export const authOptions: NextAuthOptions = {
  callbacks: {
    session({ session, user }) {
      if (session.user) {
        session.user.id = user.id;
        // session.user.role = user.role; <-- put other properties on the session here
      }
      return session;
    },
  },
  adapter: PrismaAdapter(prisma),
  providers: [
    DiscordProvider({
      clientId: env.DISCORD_CLIENT_ID,
      clientSecret: env.DISCORD_CLIENT_SECRET,
    }),
    /**
     * ...add more providers here.
     *
     * Most other providers require a bit more work than the Discord provider. For example, the
     * GitHub provider requires you to add the `refresh_token_expires_in` field to the Account
     * model. Refer to the NextAuth.js docs for the provider you want to use. Example:
     *
     * @see https://next-auth.js.org/providers/github
     */
  ],
};

/**
 * Wrapper for `getServerSession` so that you don't need to import the `authOptions` in every file.
 *
 * @see https://next-auth.js.org/configuration/nextjs
 */
export const getServerAuthSession = (opts: CreateNextContextOptions | NodeHTTPCreateContextFnOptions<IncomingMessage, ws>) => {
  const { req, res } = opts;
  // const isServer = typeof window === "undefined";
  // return !isServer ? getServerSession(req as GetServerSidePropsContext["req"], res as GetServerSidePropsContext["res"], authOptions) : getSession(opts);
  return getSession(opts);
};
```

~/utils/api.ts
```typescript
/**
 * This is the client-side entrypoint for your tRPC API. It is used to create the `api` object which
 * contains the Next.js App-wrapper, as well as your type-safe React Query hooks.
 *
 * We also create a few inference helpers for input and output types.
 */
import { httpBatchLink, loggerLink, splitLink, httpLink } from "@trpc/client";
import { wsLink, createWSClient } from "@trpc/client/links/wsLink";
import { createTRPCNext } from "@trpc/next";
import { type inferRouterInputs, type inferRouterOutputs } from "@trpc/server";
import type { NextPageContext } from "next";
import superjson from "superjson";

import { type AppRouter } from "~/server/api/root";

const headers: Map<string, string> = new Map<string, string>();

const getBaseUrl = () => {
  if (typeof window !== "undefined") return ""; // browser should use relative url
  if (process.env.VERCEL_URL) return `https://${process.env.VERCEL_URL}`; // SSR should use vercel url
  return `https://localhost:${process.env.PORT ?? 3000}`; // dev SSR should use localhost
};

const getEndingSplitLink = (ctx: NextPageContext | undefined) => {
  if (typeof window === "undefined") {
    console.log(`getEndingSplitLink gets run (server)`);
    return httpBatchLink({
      url: `${getBaseUrl()}/api/trpc`,
      headers() {
        if (ctx?.req) {
          // on ssr, forward client's headers to the server
          return {
            ...ctx.req.headers,
            "x-ssr": "1",
          };
        }
        return {};
      },
    });
  };

  const url = process.env.NEXT_PUBLIC_WS_URL || "wss://localhost:5000";
  const client = createWSClient({
    url,
    WebSocket: global.WebSocket || null,
  })

  console.log(`getEndingSplitLink gets run (webclient)`);

  return splitLink({
    condition: (op) => {
      console.log(`Condition is being checked, type: ${op.type}`);
      return op.type === 'subscription'
    },
    true: wsLink<AppRouter>({ client }),
    false: httpBatchLink({
      url: `${getBaseUrl()}/api/trpc`,
      headers() {
        console.log(`Headers being generated from httpBatchLink...`);
        return {};
      },
    }),
  })
};

/** A set of type-safe react-query hooks for your tRPC API. */
export const api = createTRPCNext<AppRouter>({
  config({ ctx }) {
    return {
      /**
       * Transformer used for data de-serialization from the server.
       *
       * @see https://trpc.io/docs/data-transformers
       */
      transformer: superjson,

      /**
       * Links used to determine request flow from client to server.
       *
       * @see https://trpc.io/docs/links
       */
      links: [
        loggerLink({
          enabled: (opts) =>
            process.env.NODE_ENV === "development" || (opts.direction === "down" && opts.result instanceof Error),
        }),
        getEndingSplitLink(ctx),
      ],
      headers() {
        return ctx?.req ? { ...ctx.req.headers } : {};
      },
      queryClientConfig: { defaultOptions: { queries: { staleTime: 60 } } },
    };
  },
  /**
   * Whether tRPC should await queries when server rendering pages.
   *
   * @see https://trpc.io/docs/nextjs#ssr-boolean-default-false
   */
  ssr: false,
});

/**
 * Inference helper for inputs.
 *
 * @example type HelloInput = RouterInputs['example']['hello']
 */
export type RouterInputs = inferRouterInputs<AppRouter>;

/**
 * Inference helper for outputs.
 *
 * @example type HelloOutput = RouterOutputs['example']['hello']
 */
export type RouterOutputs = inferRouterOutputs<AppRouter>;
```

A bit of relevant debugging information:

In the browser console I see:
`getEndingSplitLink gets run (webclient)`
` >> query #1 room.getMyRooms Object`  <-- from loggerLink showing that the http request is starting
`Condition is being checked, type: query` <-- from the splitLink condition param
`Headers being generated from httpBatchLink...` <-- from the headers function of the false branch of the splitLink

In the server console I see:
`getEndingSplitLink gets run (server)`
`API resolved without sending a response for /api/trpc/room.getMyRooms?batch=1&input=%7B%220%22%3A%7B%22json%22%3Anull%2C%22meta%22%3A%7B%22values%22%3A%5B%22undefined%22%5D%7D%7D%7D, this may result in stalled requests.`

And finally in the wsServer console, I see: (I'm using a self signed cert, so I have NODE_TLS_REJECT_UNAUTHORIZED set to 0 in env)
`Attempting to get the session from the server...`
`++ ws connection (1)`
`(node:119600) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
(Use `node-18 --trace-warnings ...` to show where the warning was created)`
`Attempting to get the session from the server...`
`++ ws connection (2)`
`createTRPCContext for User`
`createTRPCContext for User`
